tiledb/api/src/query/read/output.rs

```diff
@@ -0,0 +1,412 @@
+use std::fmt::{Debug, Formatter, Result as FmtResult};
+use std::iter::FusedIterator;
+use std::ops::{Deref, DerefMut};
+
+use anyhow::anyhow;
+
+use crate::convert::CAPISameRepr;
+use crate::error::{DatatypeErrorKind, Error};
+use crate::query::write::input::{Buffer, InputData};
+use crate::Result as TileDBResult;
+
+pub enum BufferMut<'data, T = u8> {
+    Empty,
+    Borrowed(&'data mut [T]),
+    Owned(Box<[T]>),
+}
+
+impl<'data, T> BufferMut<'data, T> {
+    pub fn size(&self) -> usize {
+        std::mem::size_of_val(self.as_ref())
+    }
+}
+
+impl<'data, T> AsRef<[T]> for BufferMut<'data, T> {
+    fn as_ref(&self) -> &[T] {
+        match self {
+            BufferMut::Empty => unsafe {
+                std::slice::from_raw_parts(
+                    std::ptr::NonNull::dangling().as_ptr(),
+                    0,
+                )
+            },
+            BufferMut::Borrowed(data) => data,
+            BufferMut::Owned(data) => data,
+        }
+    }
+}
+
+impl<'data, T> AsMut<[T]> for BufferMut<'data, T> {
+    fn as_mut(&mut self) -> &mut [T] {
+        match self {
+            BufferMut::Empty => unsafe {
+                std::slice::from_raw_parts_mut(
+                    std::ptr::NonNull::dangling().as_ptr(),
+                    0,
+                )
+            },
+            BufferMut::Borrowed(data) => data,
+            BufferMut::Owned(data) => &mut *data,
+        }
+    }
+}
+
+impl<'data, T> Deref for BufferMut<'data, T> {
+    type Target = [T];
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl<'data, T> DerefMut for BufferMut<'data, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.as_mut()
+    }
+}
+
+pub struct OutputLocation<'data, T = u8> {
+    pub data: BufferMut<'data, T>,
+    pub cell_offsets: Option<BufferMut<'data, u64>>,
+}
+
+impl<'data, T> OutputLocation<'data, T> {
+    pub fn borrow<'this>(&'this self) -> InputData<'data, T>
```

Comment:
Yes - `Borrow` returns a reference.

For now I have changed this to `as_shared` but it would be nice to have some way of "casting" a reference to `QueryBuffersMut` to `QueryBuffers` which would provide some extra utility with the `RefCell`s.